i18n: Prepare package for Calypso

sirreal opened the pull request:
See updated PR description after additional significant work: https://github.com/Automattic/wp-calypso/pull/29229#issuecomment-463233965

<details>

<summary>Original PR details</summary>

#### Changes proposed in this Pull Request

This brings the recently imported i18n-calypso package in alignment with Calypso code.

- Use packages build step
- es6 code
- Use `import` for client (compiled) code
- Fix eslint issues
- Prettier
- Run tests as part of Jest testing
- File restructuring

~This _does not_ use the package in Calypso. That can be done in a minimal, easy-to-revert PR (#29980)~ Closes #29980

#### Testing instructions

* Are tests running and passing?
* Eslint is passing?
* The code changes make sense?

</details>

sirreal committed:
Update for Calypso repo

sirreal commented on packages/i18n-calypso/src/number-format.js:
```diff
@@ -0,0 +1,38 @@
+ * @see https://github.com/kvz/phpjs/blob/ffe1356af23a6f2512c84c954dd4e828e92579fa/functions/strings/number_format.js
+ */
+
+export default function number_format( number, decimals, dec_point, thousands_sep ) {
```
This is available via npm now as part of [locutus](https://npmjs.com/package/locutus). Let's see what impact that has on the build. We might also consider pulling out this specific package in isolation if licenses allow.

At a glance, the implementation appears largely the same:

https://github.com/kvz/locutus/blob/485a9562b0435c646013bf27e84941a946ec3c97/src/php/strings/number_format.js

sirreal replied:
Let's leave this for a follow up

sirreal commented on packages/i18n-calypso/cli/README.md:
```diff
@@ -0,0 +1,39 @@
+
+## License
+
+GlotPress.js is licensed under [GNU General Public License v2 (or later)](./LICENSE.md).
```
GlotPress.js? Use an external link?

sirreal commented on packages/i18n-calypso/cli/formatters/pot.js:
```diff
@@ -0,0 +1,123 @@
+
+			return match;
+		} )
+		.filter( function( match ) {
```
`.filter( Boolean )`

sirreal commented on packages/i18n-calypso/cli/index.js:
```diff
@@ -0,0 +1,118 @@
+ */
+const fs = require( 'fs' );
+const path = require( 'path' );
+const Xgettext = require( 'xgettext-js' );
```
Next monorepofication target?

sirreal commented on packages/i18n-calypso/cli/index.js:
```diff
@@ -0,0 +1,118 @@
+/** @format */
```
Perhaps this can be dropped soon with the work on #28504 #28583 #28842, etc.

Let's not spend a lot of time on CLI cleanup.

sirreal commented on packages/i18n-calypso/cli/test/.gitignore:
```diff
@@ -0,0 +1 @@
+out
```
Is this needed? If tests make it, they should clean up.

sirreal commented on packages/i18n-calypso/cli/util.js:
```diff
@@ -0,0 +1,19 @@
+}
+
+module.exports = {
+	concatAllFiles: concatAllFiles,
```
drop repeat

sirreal commented on packages/i18n-calypso/src/i18n.js:
```diff
@@ -0,0 +1,458 @@
+	return null;
+}
+
+function I18N() {
```
This can become a real class.

sirreal commented on packages/i18n-calypso/test/index.js:
```diff
@@ -0,0 +1,346 @@
+				} );
+			} );
+
+			it( 'should make the new translations available', function() {
```
it -> test

blowery replied:
do we have a pref? `it` [is an alias](https://jestjs.io/docs/en/api.html#testname-fn-timeout) to `test` in jest.

sirreal committed:
Drop extraneous eslint extends

sirreal committed:
Disallow import from cli

sirreal committed:
Add cli/test eslint

sirreal committed:
Update license

sirreal committed:
Enable i18n tests

sirreal committed:
Ignore .eslintrc.* from test matches

sirreal committed:
Organize cli and bin

sirreal committed:
Fix const eslint error

sirreal committed:
Move expect into tests

sirreal committed:
Remove gitignore

sirreal committed:
Remove commented line

sirreal commented on packages/i18n-calypso/package.json:
```diff
@@ -2,47 +2,37 @@
-    "chai": "^3.5.0",
-    "enzyme": "^3.2.0",
-    "enzyme-adapter-react-16": "^1.1.1",
-    "enzyme-adapter-react-helper": "^1.2.3",
```
Fixes https://github.com/Automattic/wp-calypso/network/alert/packages/i18n-calypso/package-lock.json/sync-exec/closed

sirreal committed:
Cleanup

ockham commented:
If we switch Calypso to use the `@automattic/babel-plugin-i18n-calypso` and `@wordpress/babel-plugin-makepot` plugins, we might be able to remove the entire `cli/` dir and `xgettext-js` dependency here. Think it's worth trying to prioritize that? (Might be fairly easy, I'm now leaning towards a dedicated Babel config file. I might give it a quick stab.)

sirreal commented:
> Think it's worth trying to prioritize that?

I don't think we should prioritize that, I'd be concerned about getting stuck in a between state where nothing works 😬 

I'd prefer to finish this transition so the existing, relatively stable system continues to be available. After that we could move onto another system with the current one ready as a fallback.

sirreal committed:
Prettier

a8ci18n commented on packages/i18n-calypso/cli/test/examples/i18n-test-examples.jsx:
```diff
@@ -65,21 +65,23 @@ corners.` );
 
-	content = this.translate( 'The string key text',
-		{ 'context': 'context with a literal string key' } );
+	content = this.translate( 'The string key text', {
```
:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

sirreal replied:
@Automattic/i18n Is there a way to ignore strings from translation? Should we be concerned about this?

sirreal replied:
Sounds like we can just leave them, thanks 👍

lsl replied:
I'd rather them removed but you probably don't need the overhead here so: #30143

sirreal committed:
No es6 modules in cli

jsnajdr committed:
Fix the translate call in getWebPaymentMethodName

The function calls method on a `i18n` object that doesn't exist as export in `i18n-calypso`:
```js
import { i18n } from 'i18n-calypso';
```
The result is a runtime exception.

`i18n-calypso` exports the `i18n` object as a default export and the `translate` bound method
as a named export.

Fixed by passing `translate` as a second parameter. That ensures that the translations are reactive,
i.e., the React UI is rerendered when locale is changed at runtime.

Includes also a drive-by fix for translation reactivity in `PaymentBox`.

jsnajdr committed:
Expose emitter methods (on, off, emit) on the I18N class

Allows us to export a fully-functional `i18n` object as default export.

jsnajdr committed:
Convert the module exports to ESM

jsnajdr committed:
Convert the localize HOC to ES6 class

jsnajdr committed:
Update ESM imports in i18n-calypso tests

jsnajdr committed:
Update number-format source to contemporary standards

jsnajdr committed:
Improve variable declaration style in I18N, move export to ESM

Declare variables at the place of first use or inside scoped blocks, as is common
in new code. Migrate the module to full ESM.

jsnajdr committed:
Add hashbang comment to the CLI binary

jsnajdr committed:
CLI: declare variables in more contemporary ES6 style, use object property shorthands

Variables are scoped or declared at the site of first use, the way new code tends to be written.

jsnajdr committed:
Update the dependencies of i18n-calypso

Remove unused dependencies, require React 16, use one Lodash, harmonize versions
with parent package.

jsnajdr committed:
Switch Calypso to use the local i18n-calypso package

jsnajdr commented:
I rebased @sirreal's old branch onto the latest master and pushed a few commits that finish the project:
- move all exports from CommonJS to ESM. Webpack wouldn't correctly bundle a module that has both `import` and `module.exports = ` statements anyway. Webpack told me about an incorrect ESM import: #30750.
- expose event emitter methods `on`, `off` and `emit` on the I18N class. That helps make the default export (`i18n` instance) and the bound methods exported as named exports the same. Previously, `on`/`off` were named exports, but not `i18n` methods.
- convert the `localize` HOC to ES6 class instead of `create-react-class`.
- modernize the `number-format` code.
- add hashbang line at the start of `cli/index.js`. Otherwise it doesn't work (`npm run translate` fails).
- improve the way how we declare variables in a few modules. Close to first usage site, scoped, one declaration per variable. These patches originated in a funny way: I started updating the `i18n-calypso` without realizing that @sirreal's PR already exists and discovered it only halfway through. After rebasing the changes on top of each other, only a few differences remained 🙂 

**How to test:**
- verify that localized Calypso works as expected (`flags=quick-language-switcher` is your friend)
- verify that `npm run translate` produces correct results

**Roadmap for further improvements (in other PRs):**
- create a separate `i18n-calypso-cli` package with the CLI tool. There is no overlap between the client library and the tool. One of the benefits will be better separated dependencies: client library needs `dependencies`, the tool is a dev-dependency. Or maybe the CLI will be completely replaced with something Gutenbergish? @ockham and @akirk did some work recently that I didn't 100% follow.
- remove moment.js and `moment-timezone` and move them to separate libraries. Will make the Calypso bundle much smaller.
- start using React Context to update the `localize` instances. Now we use an `EventEmitter` where each instance is a subscriber. There are 100+ listeners on an average Calypso page. Maybe page navigations would be faster with Context?
- migrate from Jed to Tannin: @aduth reports performance improvements in Wordpress/gutenberg#11493

ockham commented:
> - create a separate `i18n-calypso-cli` package with the CLI tool. There is no overlap between the client library and the tool. One of the benefits will be better separated dependencies: client library needs `dependencies`, the tool is a dev-dependency. Or maybe the CLI will be completely replaced with something Gutenbergish? @ockham and @akirk did some work recently that I didn't 100% follow.

We should be able to kill the CLI altogether and replace it with @akirk's [`babel-plugin-i18n-calypso`](https://github.com/Automattic/wp-calypso/tree/master/packages/babel-plugin-i18n-calypso) once we finish and merge #29429 to use it for string extraction.

sirreal commented on package.json:
```diff
@@ -126,7 +126,7 @@
     "html-to-react": "1.3.4",
     "html-webpack-plugin": "3.2.0",
-    "i18n-calypso": "3.0.0",
+    "i18n-calypso": "file:./packages/i18n-calypso",
```
Highlighting that this _will_ put the package into production

jsnajdr replied:
Yes, this is the best part 😈

sirreal commented:
I tried rebasing this on master, running `npm run translate` on master and on this branch, and comparing the results in `calypso-strings.pot`.

The results are very different. I'm not sure whether that's always been an issue or if the results should be relatively stable. I'd appreciate if someone from @Automattic/i18n could take a look to see whether the output is as expected.

I've been smoke testing in Calypso using the fast language switcher and that i18n/translation part of this seems to work well 👍

jsnajdr commented:
> The results are very different. I'm not sure whether that's always been an issue or if the results should be relatively stable.

Looking at the diff of both files in FileMerge, it seems that the order of things is different. Otherwise the content is the same. I'll try to figure out what's going on.

jsnajdr commented:
> I'll try to figure out what's going on.

I think it's because `globby` got upgraded from 6.x to 9.x. Especially the 8.0.0 upgrade [switched the underlying globbing library](https://github.com/sindresorhus/globby/releases/tag/v8.0.0).

The old version lists the matching files in `a` first, then in `b`, including subdirectories:
```
a/file.js
a/a/file.s
a/a/a/file.js
b/file.js
b/b/file.js
b/b/b/file.js
```

The new version lists the depth-1 directories first, the the depth-2 ones, etc.:
```
a/file.js
b/file.js
a/a/file.s
b/b/file.js
a/a/a/file.js
b/b/b/file.js
```

That's all 🙂

blowery committed:
add a comment to the lint rules for the i18n cli to explain why

blowery commented on packages/i18n-calypso/cli/.eslintrc.js:
```diff
@@ -0,0 +1,10 @@
+
+module.exports = {
+	parserOptions: {
+		sourceType: 'script',
```
added a comment to that effect

blowery commented on packages/i18n-calypso/package.json:
```diff
@@ -1,48 +1,37 @@
   "description": "i18n JavaScript library on top of Jed originally used in Calypso",
-  "main": "index.js",
+  "main": "dist/cjs/index.js",
+  "module": "dist/esm/index.js",
```
`sideEffects: false` should be useful here too. I don't think we have any side effects in index?

blowery replied:
added in https://github.com/Automattic/wp-calypso/pull/29229/commits/911996ecbcbcc655ac7743369b40b4b14c7e3fe3

jsnajdr replied:
I don't think that `sideEffects` will have much effect here: all we export is parts of the `i18n` object instance. The only thing that one can shake off is the `localize` HOC. But it doesn't do any harm.

blowery commented:
I noticed that `i18n-calypso` had a `node_modules` in it after the `npm ci`, which was surprising. `globby` was there, but the wrong version... `npm ls globby` showed it as invalid.

`npm i --package-lock-only` cleared that up, replacing globby 8.0.2 with 9.0.0. 
`npm dedupe` then deduped the rest of the dependencies down to the root.

blowery committed:
add sideEffects: false to help tree-shaking

blowery commented:
This all looks good to me! I have some code in here now, so I'd appreciate a review on those little bits.

blowery commented:
@jsnajdr @sirreal rebased this on current master and force pushed

jsnajdr commented:
I like and approve all your changes @blowery. I also rebased onto the latest master to resolve the package lock conflicts.

lsl commented:
> I'd appreciate if someone from @Automattic/i18n could take a look to see whether the output is as expected.

Guessing the rebase resolved the issue you were seeing at the time of that comment. Current (rebased) output vs master LGTM. Only seeing the expected line ref changes.

sirreal committed:
Update and dedupe deps

This commit is the result of running:

npm run update-deps
npm dedupe

sirreal commented:
I've rebased on latest master and force pushed.

I've removed conflicting shrinkwrap changes from the middle and stacked them in a single commit at the HEAD:

```
npm run update-deps
npm dedupe
```

jsnajdr reviewed (approved):
I believe this PR is ready for merging, isn't it? @blowery and @sirreal what do you think?

sirreal commented:
> I believe this PR is ready for merging, isn't it? @blowery and @sirreal what do you think?

I was hoping for the full e2e suite to pass before landing this, but the test suite isn't reliable enough for branches.

This is working well in my testing, I'd approve but I'm the PR author 🙂 

✅ from me 👍

jsnajdr commented:
> I was hoping for the full e2e suite to pass before landing this

Sure thing, I'm surprised we didn't run that already 🙂 Let's also wait what Ben (the third coauthor) has to say.

blowery committed:
move shrinkwrap back to tabs to cut down the diff

blowery reviewed (approved):
LGTM.

Just moved the shrinkwrap back to using tabs, which cut down on the diff size considerably.
